main/com.spotify/index.js

```diff
@@ -523,23 +533,52 @@ module.exports = class SpotifyDevice extends Tp.BaseDevice {
         });
     }
 
+    async get_available_device_ids() {
+        return this.http_get(AVAILABLE_DEVICES_URL).then((response) => {
+            const devices = JSON.parse(response).devices;
+            return devices.map((available_device) => available_device.id);
+        });
+    }
+
     _testMode() {
         return process.env.TEST_MODE === '1';
     }
 
-    _findActiveDevice(devices) {
-        if (devices.length === 0) {
-            console.log("no available devices");
-            return [null, null];
-        }
+    async _findActiveDevice(devices) {
         // device already active
+        let spotifydActive = false;
         for (let i = 0; i < devices.length; i++) {
-            console.log(devices[i].is_active);
+            if (devices[i].id === this._deviceId)
+                spotifydActive = true;
+
             if (devices[i].is_active) {
                 console.log("found an active device");
                 return [devices[i].id, devices[i].name];
             }
         }
+        if (spotifydActive) {
+            return [this._deviceId, this.userId];
+        } else if (this._deviceId) {
+            //this code attempts to restart a spotifyd instance
+            const code = `killall spotifyd; spotifyd --username ${this.state.profile.id} --device-name ${this.state.profile.id} --token ${this.accessToken}`;
+            exec(code, (error, stdout, stderr) => {
+                if (error) throw new Error(error);
+                if (stderr) throw new Error(stderr);
+            });
+            let deviceIds = await this.get_available_device_ids();
+            const startTime = new Date().getTime();
+            //it takes a couple seconds for spotify's api to detect the new device
+            while (new Date().getTime() - startTime < 10000) {
```

Comment:
This is not how you do a timeout. Instead, you want someting like:
```
function delay(ms) {
    return new Promise((resolve) => {
        setTimeout(resolve, ms);
    });
}
await delay(1000);
```
in a loop.

Otherwise, you hammer the API (burning through the quota) and also waste CPU cycles.